From d0ea6a3e4349184b9b6cadf8ff8f636cb7e6c352 Mon Sep 17 00:00:00 2001
From: Zeng Heng <zengheng4@huawei.com>
Date: Wed, 23 Apr 2025 18:51:41 +0800
Subject: [PATCH 698/960] downstream: arm64/mpam: Optimize CSU/MBWU monitor
 multiplexing

When MBM is implemented according to the *DDI0598* protocol, that is,
mbm_total_bytes interface returns the statistical results of instantaneous
bandwidth traffic, there is no need to create a delay_work to handle MBM
statistical overflow.

No matter it is MBWU or CSU monitors, using per-CLOSID monitoring instances
avoids inaccurate statistical results caused by monitor multiplexing, which
can occur when multiple control groups are monitored simultaneously.

Fixes: 9119da143939 ("arm_mpam: resctrl: Allow resctrl to allocate monitors")
Signed-off-by: Zeng Heng <zengheng4@huawei.com>
Signed-off-by: Liu Shixin <liushixin2@huawei.com>
---
 drivers/platform/mpam/mpam_resctrl.c | 33 +++++++---------------------
 1 file changed, 8 insertions(+), 25 deletions(-)

diff --git a/drivers/platform/mpam/mpam_resctrl.c b/drivers/platform/mpam/mpam_resctrl.c
index 7cdba4516e16..e3acda04bd3a 100644
--- a/drivers/platform/mpam/mpam_resctrl.c
+++ b/drivers/platform/mpam/mpam_resctrl.c
@@ -254,7 +254,6 @@ struct rdt_resource *resctrl_arch_get_resource(enum resctrl_res_level l)
 static void *resctrl_arch_mon_ctx_alloc_no_wait(struct rdt_resource *r,
 						int evtid)
 {
-	struct mpam_resctrl_res *res;
 	u32 *ret = kmalloc(sizeof(*ret), GFP_KERNEL);
 
 	if (!ret)
@@ -262,11 +261,6 @@ static void *resctrl_arch_mon_ctx_alloc_no_wait(struct rdt_resource *r,
 
 	switch (evtid) {
 	case QOS_L3_OCCUP_EVENT_ID:
-		res = container_of(r, struct mpam_resctrl_res, resctrl_res);
-
-		*ret = mpam_alloc_csu_mon(res->class);
-		return ret;
-
 	case QOS_L3_MBM_LOCAL_EVENT_ID:
 	case QOS_L3_MBM_TOTAL_EVENT_ID:
 		*ret = __mon_is_rmid_idx;
@@ -300,21 +294,7 @@ void *resctrl_arch_mon_ctx_alloc(struct rdt_resource *r, int evtid)
 void resctrl_arch_mon_ctx_free(struct rdt_resource *r, int evtid,
 			       void *arch_mon_ctx)
 {
-	struct mpam_resctrl_res *res;
-	u32 mon = *(u32 *)arch_mon_ctx;
-
 	kfree(arch_mon_ctx);
-
-	switch (evtid) {
-	case QOS_L3_OCCUP_EVENT_ID:
-		res = container_of(r, struct mpam_resctrl_res, resctrl_res);
-		mpam_free_csu_mon(res->class, mon);
-		wake_up(&resctrl_mon_ctx_waiters);
-		return;
-	case QOS_L3_MBM_TOTAL_EVENT_ID:
-	case QOS_L3_MBM_LOCAL_EVENT_ID:
-		return;
-	}
 }
 
 static enum mon_filter_options resctrl_evt_config_to_mpam(u32 local_evt_cfg)
@@ -335,7 +315,7 @@ int resctrl_arch_rmid_read(struct rdt_resource	*r, struct rdt_domain *d,
 {
 	int err;
 	u64 cdp_val;
-	u16 num_mbwu_mon;
+	u16 num_mon;
 	struct mon_cfg cfg;
 	struct mpam_resctrl_dom *dom;
 	struct mpam_resctrl_res *res;
@@ -362,12 +342,15 @@ int resctrl_arch_rmid_read(struct rdt_resource	*r, struct rdt_domain *d,
 	if (cfg.mon == USE_RMID_IDX) {
 		/*
 		 * The number of mbwu monitors can't support free run mode,
-		 * adapt the remainder of rmid to the num_mbwu_mon as a
-		 * compromise.
+		 * adapt the remainder of rmid to the num_mon as compromise.
 		 */
 		res = container_of(r, struct mpam_resctrl_res, resctrl_res);
-		num_mbwu_mon = res->class->props.num_mbwu_mon;
-		cfg.mon = resctrl_arch_rmid_idx_encode(closid, rmid) % num_mbwu_mon;
+		if (type == mpam_feat_msmon_mbwu)
+			num_mon = res->class->props.num_mbwu_mon;
+		else
+			num_mon = res->class->props.num_csu_mon;
+
+		cfg.mon = closid % num_mon;
 	}
 
 	cfg.match_pmg = true;
-- 
2.43.0

